[TCP] Support for tcp.cast op

zezhang opened the pull request:
As titled.

Testing:

```
./build/bin/llvm-lit externals/llvm-external-projects/torch-mlir-dialects/test/Conversion/TcpToLinalg/ -v
./build/bin/llvm-lit externals/llvm-external-projects/torch-mlir-dialects/test/Dialect/Tcp/ -v
./build/bin/llvm-lit test/Conversion/TorchToTcp/ -v
python -m e2e_testing.main --config=tcp -v -s
```

navahgar commented on externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td:
```diff
@@ -514,4 +514,28 @@ def Tcp_Atan2Op : Tcp_BinaryElementwiseOp<"atan2", [SameOperandsAndResultElement
+
+  let arguments = (ins
+    Tcp_Tensor:$in,
+    // Align with enum SignednessSemantics
```
Is there a better way to handle attributes that are enum?

navahgar replied:
@burmako What is the recommended approach for defining attributes that represent enums like this?

burmako replied:
Here's how we're doing this in StableHLO:
  * Definition of an enum attribute: https://github.com/openxla/stablehlo/blob/d67da5d85045e03893684c35ae5e730695289ae2/stablehlo/dialect/StablehloEnums.td#L98-L122.
  * Use of this attribute: https://github.com/openxla/stablehlo/blob/d67da5d85045e03893684c35ae5e730695289ae2/stablehlo/dialect/StablehloOps.td#L1570.

navahgar replied:
Thanks a lot for those pointers @burmako. Thats very helpful.

@zezhang Can we follow a similar approach to represent signedness in `tcp.cast`?

navahgar commented on externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td:
```diff
@@ -514,4 +514,28 @@ def Tcp_Atan2Op : Tcp_BinaryElementwiseOp<"atan2", [SameOperandsAndResultElement
+  let arguments = (ins
+    Tcp_Tensor:$in,
+    // Align with enum SignednessSemantics
+    OptionalAttr<UI32Attr>:$in_dtype,
+    OptionalAttr<UI32Attr>:$out_dtype
```
Lets not call these `dtype`. I believe MLIR uses `elementType` instead.

Also, we should probably encode in the name that this is only meant for integer types.

navahgar commented on externals/llvm-external-projects/torch-mlir-dialects/lib/Conversion/TcpToLinalg/Elementwise.cpp:
```diff
@@ -174,24 +174,108 @@ createLinalgPayloadForElementwiseOp(Operation *op,
+            outputType.getIntOrFloatBitWidth() &&
+        ((!castOp.getInDtype() && !castOp.getOutDtype()) ||
+         (castOp.getInDtype() && castOp.getOutDtype() &&
+          castOp.getInDtype().value() == castOp.getOutDtype().value())))
```
nit: since you are adding `{ }` for the else cases below, it is better to add that here as well, to keep it uniform and more readable.

navahgar commented on externals/llvm-external-projects/torch-mlir-dialects/lib/Conversion/TcpToLinalg/Elementwise.cpp:
```diff
@@ -174,24 +174,108 @@ createLinalgPayloadForElementwiseOp(Operation *op,
+         (castOp.getInDtype() && castOp.getOutDtype() &&
+          castOp.getInDtype().value() == castOp.getOutDtype().value())))
+      return payloadArgs[0];
+    // To I1 (Bool)
```
nit: To make it more readable, consider formatting this like:

```
   } else if (outputType.isInteger(1)) { // to I1 (Bool)
```
OR
```
   } else if (outputType.isInteger(1)) {
      // to I1 (Bool)
```

navahgar commented on externals/llvm-external-projects/torch-mlir-dialects/lib/Conversion/TcpToLinalg/Elementwise.cpp:
```diff
@@ -174,24 +174,108 @@ createLinalgPayloadForElementwiseOp(Operation *op,
+      }
+    }
+    // TO FP
+    else if (!castOp.getOutDtype()) {
```
It is better to use the condition `outputType.isa<mlir::FloatType>()` here instead.

navahgar commented on externals/llvm-external-projects/torch-mlir-dialects/lib/Conversion/TcpToLinalg/Elementwise.cpp:
```diff
@@ -174,24 +174,108 @@ createLinalgPayloadForElementwiseOp(Operation *op,
+      }
+    }
+    // TO INT
+    else if (castOp.getOutDtype()) {
```
same as above

navahgar commented on lib/Conversion/TorchToTcp/Elementwise.cpp:
```diff
@@ -417,13 +417,207 @@ class ConvertAtenAtan2Op : public OpConversionPattern<AtenAtan2Op> {
+    } else if (nonBlocking) {
+      return rewriter.notifyMatchFailure(
+          op, "unimplemented: non_blocking arg is expected to be false");
+    }
```
nit: To simplify the code, we can probably combine the two checks into one.

navahgar commented on lib/Conversion/TorchToTcp/Elementwise.cpp:
```diff
@@ -417,13 +417,207 @@ class ConvertAtenAtan2Op : public OpConversionPattern<AtenAtan2Op> {
+    } else if (copy) {
+      return rewriter.notifyMatchFailure(
+          op, "unimplemented: copy arg is expected to be false");
+    }
```
same as above

navahgar commented on lib/Conversion/TorchToTcp/Elementwise.cpp:
```diff
@@ -417,13 +417,207 @@ class ConvertAtenAtan2Op : public OpConversionPattern<AtenAtan2Op> {
+    // Only `none`, `contiguous` and `preserve` memory_format is supported.
+    if (!op.getMemoryFormat().getType().isa<Torch::NoneType>()) {
+      int64_t memoryFormat;
+      if (!matchPattern(op.getMemoryFormat(),
+                        m_TorchConstantInt(&memoryFormat)))
+        return rewriter.notifyMatchFailure(
+            op, "unimplemented: the memory format should be specified in "
+                "an integer constant");
+      if (memoryFormat != torch_upstream::MemoryFormat::Contiguous &&
+          memoryFormat != torch_upstream::MemoryFormat::Preserve)
+        return rewriter.notifyMatchFailure(
+            op, "unimplemented: only none, contiguous and preserve "
+                "memory_format is supported");
```
same as above

navahgar commented on lib/Conversion/TorchToTcp/Elementwise.cpp:
```diff
@@ -417,13 +417,207 @@ class ConvertAtenAtan2Op : public OpConversionPattern<AtenAtan2Op> {
+          op, resultType, adaptor.getSelf(), IntegerAttr{}, IntegerAttr{});
+    // FP -> INT
+    else if (inputType.getDtype().isa<mlir::FloatType>()) {
+      if (outputType.getDtype().isSignlessInteger())
```
Instead of checking for these conditions, you can directly use the method `getSignedness()` to get the signedness of the input and output types. That would simplify this and the code below a lot.

navahgar reviewed (commented):
Took one pass over it.

navahgar commented on externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td:
```diff
@@ -514,4 +514,28 @@ def Tcp_Atan2Op : Tcp_BinaryElementwiseOp<"atan2", [SameOperandsAndResultElement
+  let arguments = (ins
+    Tcp_Tensor:$in,
+    // TODO: Better way to align with enum SignednessSemantics
+    OptionalAttr<UI32Attr>:$in_int_element_type,
```
If this is only meant to represent the `SignednessSemantics`, we should probably name it accordingly, say something like `in_int_signedness`. 

My earlier comment was just that `dtype` is not a term used in core MLIR.

navahgar commented on externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td:
```diff
@@ -514,4 +514,28 @@ def Tcp_Atan2Op : Tcp_BinaryElementwiseOp<"atan2", [SameOperandsAndResultElement
+  let summary = "TCP Cast operation";
+
+  let description = [{
+    Performs elementwise type conversion.
```
Can you add more explanation as to the type of casts we do here?

This is probably a good place to document all casting cases that are covered.

navahgar reviewed (commented):
Mostly LGTM. Just a couple of minor comments to address.

navahgar commented:
@burmako Can you please review this PR?

navahgar reviewed (commented):
LGTM. Thanks for taking this up. I don't have any other comments. 

I will let @burmako approve the PR.

burmako commented on externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td:
```diff
@@ -514,4 +515,41 @@ def Tcp_Atan2Op : Tcp_BinaryElementwiseOp<"atan2", [SameOperandsAndResultElement
+      FP -> FP
+      INT -> FP
+      FP -> INT
+      INT -> INT
```
There are quite a few corner cases here, e.g. what happens if the conversion overflows the destination type or what happens if a negative number is converted to an unsigned destination type. Consider explicitly speccing them at some point.

navahgar replied:
Thats a good point. @zezhang Can you add a TODO here to handle other corner cases?

burmako commented on externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td:
```diff
@@ -514,4 +515,41 @@ def Tcp_Atan2Op : Tcp_BinaryElementwiseOp<"atan2", [SameOperandsAndResultElement
+  let arguments = (ins
+    Tcp_Tensor:$in,
+    OptionalAttr<Tcp_SignednessAttr>:$in_int_signedness,
+    OptionalAttr<Tcp_SignednessAttr>:$out_int_signedness
```
Have you considered an alternative design with multiple ops like in the Arith dialect? I.e. `sitofp`, `uitofp`, `fptosi`, `fptoui`, `extsi`, `extui`, `extf`, `trunci`, `truncf`?

navahgar replied:
We debated about this. The summary was that we felt that is probably too low level of an abstraction for TCP. And it is better to represent "casting" as a single op so that transformations are simpler.

Let us know your thoughts on this. Are we missing anything obvious here?

burmako replied:
I can't say I have hard data on what approach should be preferred in this case. The optional signedness attributes felt a bit awkward, but having a bunch of ops instead of just one also has downsides. I think that the current design looks good for getting started and getting more experience with it.

navahgar replied:
Sounds good. Lets keep this as one of the design points that needs to be revisited.

burmako commented on externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp:
```diff
@@ -155,4 +156,43 @@ LogicalResult ConcatOp::verify() {
+                         "Signedness::Signless when output is i1");
+  }
+
+  return success();
```
What if inputType and/or outputType aren't integer or floating-point tensors? As it's currently written, the verifier will succeed but that's probably undesirable?

burmako commented on externals/llvm-external-projects/torch-mlir-dialects/lib/Conversion/TcpToLinalg/Elementwise.cpp:
```diff
@@ -174,24 +174,97 @@ createLinalgPayloadForElementwiseOp(Operation *op,
+      else if (inputType.dyn_cast<mlir::IntegerType>()) {
+        // Signless or Unsigned INT to FP
+        if (castOp.getInIntSignedness().value() == Signedness::Signless ||
+            castOp.getInIntSignedness().value() == Signedness::Unsigned)
```
Why does signless mean unsigned? This was a bit unexpected.

navahgar replied:
IIRC, this bit is meant for `i1` (`bool`) case only. Do you see any issues with that?

@zezhang Can you document this in the op? Also, it would be good to add a check in the verifier to ensure that it is `signless` only when the integer width is 1.

burmako replied:
"IIRC, this bit is meant for i1 (bool) case only. Do you see any issues with that?". Ah, this was non-obvious :) Additional documentation would help with this.

burmako reviewed (commented):
Reviewed the enum attributes part + left a few design questions. Please feel free to ignore those - they are mostly for my curiousity.

zezhang commented:
Thanks @burmako for reviewing this. I have addressed your comments. Could you take another look and approve it if everything looks good for now. Please let me know if you have any other questions.

navahgar commented on externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td:
```diff
@@ -514,4 +515,41 @@ def Tcp_Atan2Op : Tcp_BinaryElementwiseOp<"atan2", [SameOperandsAndResultElement
+  let description = [{
+    Performs elementwise type conversion.
+
+    Supported torch/mlir element type:
```
TCP should adhere to the core types in MLIR. The comment is just misleading here. 

@zezhang Can you fix the comment to talk only about MLIR core types?

burmako commented on externals/llvm-external-projects/torch-mlir-dialects/test/Dialect/Tcp/unary.mlir:
```diff
@@ -183,3 +183,63 @@ func.func @test_atan_f32(%arg0 : tensor<?x?xf32>) -> tensor<?x?xf32> {
+  // expected-error@+1{{tcp.cast' op out_int_signedness attr must be set when output is INT}}
+  %0 = tcp.cast %arg0 : tensor<?x?xf32> -> tensor<?x?xi16>
+  return %0 : tensor<?x?xi16>
+}
```
Consider adding more negative tests to cover all the ways how verification can fail. Apologies - I didn't notice this in the previous review.

burmako reviewed (approved):
Overall, the PR looks good, and I left just one additional comment.
